### ISSUE TITLE
Bug report: need to restart to see explanaiton about the string's size limit

### ISSUE THREAD

**Issue by Arthur-Milchior** [ankidroid/Anki-Android#4731]
###### Research
*Enter an [ x ] character to confirm the points below:*

[x] I have read the [support page](https://ankidroid.org/docs/help.html) and am reporting a bug or enhancement request specific to AnkiDroid

[ x] I have checked the [manual](https://ankidroid.org/docs/manual.html) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue

[ x] I have searched for similar existing issues here and on the user forum

###### Reproduction Steps

1. Have too many note type on ankiweb
2. sync

###### Expected Result
An error message stating that Andoid's limit for string size caused a problem.
This is the message I obtain when I restart ankidroid. And I think it would be helpful to have when the sync ended.

###### Actual Result
An error message which does not really explain what is wrong, and how to solve the problem, apart from trying a new sync/using an older backup

###### Debug info
Refer to the [support page](https://ankidroid.org/docs/help.html) if you are unsure where to get the "debug info".

With this trouble, I can't access Debug info. Which should potentially be the subject of another feature request/bug report. (By the way, it would also be nice to be able to access options even with this bug. This would allow me to switch to English and give you the exact error message I have)

**Comment by timrae** [comment]
Currently we are just [checking](https://github.com/ankidroid/Anki-Android/blob/master/AnkiDroid/src/main/java/com/ichi2/anki/CollectionHelper.java#L124) if there is an `IllegalStateException` while opening the collection (after the sync has completed), and if there is then we say that it's caused by the string issue, as that's the only issue that has caused that behavior so far.

There are several ways that we could potentially improve the reporting on this error, but I would far prefer to fix the root cause of the issue via a change in the db schema.

@dae 
Any ETA on when you could feasibly fix this issue upstream? When we previously discussed it 2 years ago, you mentioned that you'd eventually fix it by separating each model into their own cell instead of storing the raw JSON array. This issue does pop up from time to time and it usually causes a couple of hours of work for me each time, so I'd like to see it get resolved.

**Comment by Arthur-Milchior** [comment]
For you information, I just downloaded 
https://github.com/ankidroid/Anki-Android/releases/tag/v2.8.3 and the 
problem did dissapear. Even better, it didn't even ask me to do a new 
download of the database.

As far as I'm concerned, this can be closed, even if I guess that this 
kind of error is still potentially relevant (I mean, I'd really like to 
have a lot of note type, for my usage)

Regards,

Arthur Milchior


Le 11/11/2017 à 02:36, Tim Rae a écrit :

**Comment by Arthur-Milchior** [comment]
Since it seems that the problem does not really depends on the size of the field, what could I possibly do to troubleshoot this error ? Or should I quit the idea of using ankidroid ? 
I know java but don't know how to compile code for android, so I'd prefer not to have to do it if I can avoid; even if I really love this software


N.b.
My previous message was too quick, I did delete it, because the problem remained.

**Comment by dae** [comment]
Tim: because it requires changes that will break existing clients, it's on hold until a major update. I'd hoped to have started on that earlier, but Qt's awful webkit->webengine migration has meant 2.1 has taken a lot longer than I would have liked.

I suspect only a very small subset of users have enough note type content to exceed the 1MB limit, so I'm not sure it's easy to justify anything more elaborate than a useful error message for now. But if you wanted to try to fix it, perhaps it would be possible to use the length() and substr() sql functions to extract the models field in chunks, and write it back in chunks with the string concat operator.

**Comment by timrae** [comment]
@Arthur-Milchior 

As I wrote on the other thread, send me your collection

**Comment by Arthur-Milchior** [comment]
I just exported my collection (without media), and uploaded it on https://drive.google.com/open?id=1DYDGCxUEGA7_yypwnDg8A7lnQTMZuzeX
One strong warning: do not try to do check media. Because, there are 14000 latex's image, compiling them would be really long.
I didn't see your offer in the other thread. Thank you for accepting to take time to have a look at it.
Let me know if I can do anything to help. (Note, however, that this problem seems to be unrelated to the bug I reporterd in my OP. So this sounds like off-topic talk)

**Comment by timrae** [comment]
Sorry I'm a bit lost, what is unrelated to what?

**Comment by Arthur-Milchior** [comment]
My bug report was about a small fact:

-The error message arrive too late. The first error message is not 
informative. The informative mesage, stating that one should delete some 
note types, is shown only after that Ankidroid is restarted.


-The error we are currently discussing is about the unknown database 
error which occurs when I try to synchronize my database.


Both of those errors seems to be unrelated, especially since my problem 
is not really about strings which are too long.


I hope that it is more clear no.


Thanks,

Arthur


Le 11/11/2017 à 23:32, Tim Rae a écrit :
> Sorry I'm a bit lost, what is unrelated to what?
>

**Comment by timrae** [comment]
Ok I've at least reproduced that I can't open this file in ankidroid. Can
you confirm that it does in fact open with anki desktop?

**Comment by timrae** [comment]
On my Android device `libsqlite3x.so` just segfaults... no java stacktrace or anything :-/  It's also opening fine in Anki Desktop.

**Comment by timrae** [comment]
I double checked that your strings are less than the limit. Could you please follow the instructions here for corrupt collections and see if that helps?

https://apps.ankiweb.net/docs/manual.html#corrupt-collections

**Comment by Arthur-Milchior** [comment]
It didn't change anything.

More details:

No error during sqlite3 collection.anki2 .dump > dump.txt

(I did it on the file I sent you)

Last line was commit


No error during

cat dump.txt | sqlite3 temp.file
the file is not empty

I renamed collection.anki2 to collection.anki2bak
I renamed temp.fil to collection.anki2 (in .local/share/Anki2/Arthur Milchior )


I oppened anki
I check the database.
I synced on the desktop.
No error at all.
I synced on ankidroid.
Still the same error message: «La collection n'est pas ouverte
Une errer s'est produite en accédant à votre collection. ...»




Le 12/11/2017 à 15:26, Tim Rae a écrit :

**Comment by timrae** [comment]
Can you please post your logcat? I want to see the error string inside your `IllegalStateException`.  The best way to get the logcat is to set up a dev environment with Android Studio, as this will also allow you attach the debugger later if necessary. Check out these instructions and work your way down to the part about connecting over adb:

https://github.com/ankidroid/Anki-Android/wiki/Development-Guide

**Comment by Arthur-Milchior** [comment]
I finally succedeed in following all of the instructions. 
I gues that what you want is: 
11-30 15:12:34.582 698-698/? E/AnkiDroid: CollectionHelper/ Error opening collection; closing database
                                          java.lang.IllegalStateException: Couldn't read row
                                              at io.requery.android.database.CursorWindow.nativeGetLong(Native Method)
                                              at io.requery.android.database.CursorWindow.getLong(CursorWindow.java:339)
                                              at io.requery.android.database.AbstractWindowedCursor.getLong(AbstractWindowedCursor.java:80)
                                              at com.ichi2.libanki.Collection.load(Collection.java:190)
                                              at com.ichi2.libanki.Collection.<init>(Collection.java:151)
                                              at com.ichi2.libanki.Storage.Collection(Storage.java:63)
                                              at com.ichi2.anki.CollectionHelper.getCol(CollectionHelper.java:99)
                                              at com.ichi2.anki.CollectionHelper.getColSafe(CollectionHelper.java:112)
                                              at com.ichi2.anki.DeckPicker.firstCollectionOpen(DeckPicker.java:460)
                                              at com.ichi2.anki.DeckPicker.onCreate(DeckPicker.java:361)
                                              at android.app.Activity.performCreate(Activity.java:5451)
                                              at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1093)
                                              at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2377)
                                              at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2471)
                                              at android.app.ActivityThread.access$900(ActivityThread.java:175)
                                              at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1308)
                                              at android.os.Handler.dispatchMessage(Handler.java:102)
                                              at android.os.Looper.loop(Looper.java:146)
                                              at android.app.ActivityThread.main(ActivityThread.java:5602)
                                              at java.lang.reflect.Method.invokeNative(Native Method)
                                              at java.lang.reflect.Method.invoke(Method.java:515)
                                              at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1283)
                                              at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1099)
                                              at dalvik.system.NativeStart.main(Native Method)

Do you need anything else ?

**Comment by timrae** [comment]
Would you be able to downgrade to [v2.7](https://github.com/ankidroid/Anki-Android/releases/tag/v2.7) and get that same stacktrace there please?

**Comment by Arthur-Milchior** [comment]
12-01 03:22:15.550 16257-16257/? E/AnkiDroid: CollectionHelper/ Error 
opening collection; closing database
java.lang.IllegalStateException: Couldn't read row 0, col 0 from 
CursorWindow.  Make sure the Cursor is initialized correctly before 
accessing data from it.
                                                   at 
android.database.CursorWindow.nativeGetLong(Native Method)
                                                   at 
android.database.CursorWindow.getLong(CursorWindow.java:512)
                                                   at 
android.database.AbstractWindowedCursor.getLong(AbstractWindowedCursor.java:75)
                                                   at 
com.ichi2.libanki.Collection.load(Collection.java:190)
                                                   at 
com.ichi2.libanki.Collection.<init>(Collection.java:151)
                                                   at 
com.ichi2.libanki.Storage.Collection(Storage.java:63)
                                                   at 
com.ichi2.anki.CollectionHelper.getCol(CollectionHelper.java:99)
                                                   at 
com.ichi2.anki.CollectionHelper.getColSafe(CollectionHelper.java:112)
                                                   at 
com.ichi2.anki.DeckPicker.firstCollectionOpen(DeckPicker.java:461)
                                                   at 
com.ichi2.anki.DeckPicker.onCreate(DeckPicker.java:362)
                                                   at 
android.app.Activity.performCreate(Activity.java:5451)
                                                   at 
android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1093)
                                                   at 
android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2377)
                                                   at 
android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2471)
                                                   at 
android.app.ActivityThread.access$900(ActivityThread.java:175)
                                                   at 
android.app.ActivityThread$H.handleMessage(ActivityThread.java:1308)
                                                   at 
android.os.Handler.dispatchMessage(Handler.java:102)
                                                   at 
android.os.Looper.loop(Looper.java:146)
                                                   at 
android.app.ActivityThread.main(ActivityThread.java:5602)
                                                   at 
java.lang.reflect.Method.invokeNative(Native Method)
                                                   at 
java.lang.reflect.Method.invoke(Method.java:515)
                                                   at 
com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1283)
                                                   at 
com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1099)
                                                   at 
dalvik.system.NativeStart.main(Native Method)
12-01 03:22:17.222 16257-16257/? E/AnkiDroid: CollectionHelper/ Error 
opening collection; closing database
java.lang.IllegalStateException: Couldn't read row 0, col 0 from 
CursorWindow.  Make sure the Cursor is initialized correctly before 
accessing data from it.
                                                   at 
android.database.CursorWindow.nativeGetLong(Native Method)
                                                   at 
android.database.CursorWindow.getLong(CursorWindow.java:512)
                                                   at 
android.database.AbstractWindowedCursor.getLong(AbstractWindowedCursor.java:75)
                                                   at 
com.ichi2.libanki.Collection.load(Collection.java:190)
                                                   at 
com.ichi2.libanki.Collection.<init>(Collection.java:151)
                                                   at 
com.ichi2.libanki.Storage.Collection(Storage.java:63)
                                                   at 
com.ichi2.anki.CollectionHelper.getCol(CollectionHelper.java:99)
                                                   at 
com.ichi2.anki.CollectionHelper.exceededCursorSizeLimit(CollectionHelper.java:126)
                                                   at 
com.ichi2.anki.DeckPicker.onCreate(DeckPicker.java:446)
                                                   at 
android.app.Activity.performCreate(Activity.java:5451)
                                                   at 
android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1093)
                                                   at 
android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2377)
                                                   at 
android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2471)
                                                   at 
android.app.ActivityThread.access$900(ActivityThread.java:175)
                                                   at 
android.app.ActivityThread$H.handleMessage(ActivityThread.java:1308)
                                                   at 
android.os.Handler.dispatchMessage(Handler.java:102)
                                                   at 
android.os.Looper.loop(Looper.java:146)
                                                   at 
android.app.ActivityThread.main(ActivityThread.java:5602)
                                                   at 
java.lang.reflect.Method.invokeNative(Native Method)
                                                   at 
java.lang.reflect.Method.invoke(Method.java:515)
                                                   at 
com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1283)
                                                   at 
com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1099)
                                                   at 
dalvik.system.NativeStart.main(Native Method)
12-01 03:22:17.622 16257-16385/? D/ACRA: Creating DialogIntent for 
1512094936000.stacktrace exception=java.lang.IllegalStateException: 
Couldn't read row 0, col 0 from CursorWindow.  Make sure the Cursor is 
initialized correctly before accessing data from it.
12-01 03:22:17.783 16257-16257/? E/AnkiDroid: CollectionHelper/ Error 
opening collection; closing database
java.lang.IllegalStateException: Couldn't read row 0, col 0 from 
CursorWindow.  Make sure the Cursor is initialized correctly before 
accessing data from it.
                                                   at 
android.database.CursorWindow.nativeGetLong(Native Method)
                                                   at 
android.database.CursorWindow.getLong(CursorWindow.java:512)
                                                   at 
android.database.AbstractWindowedCursor.getLong(AbstractWindowedCursor.java:75)
                                                   at 
com.ichi2.libanki.Collection.load(Collection.java:190)
                                                   at 
com.ichi2.libanki.Collection.<init>(Collection.java:151)
                                                   at 
com.ichi2.libanki.Storage.Collection(Storage.java:63)
                                                   at 
com.ichi2.anki.CollectionHelper.getCol(CollectionHelper.java:99)
                                                   at 
com.ichi2.anki.CollectionHelper.getColSafe(CollectionHelper.java:112)
                                                   at 
com.ichi2.anki.DeckPicker.onPrepareOptionsMenu(DeckPicker.java:523)
                                                   at 
android.app.Activity.onPreparePanel(Activity.java:2664)
                                                   at 
android.support.v4.app.FragmentActivity.onPrepareOptionsPanel(FragmentActivity.java:547)
                                                   at 
android.support.v4.app.FragmentActivity.onPreparePanel(FragmentActivity.java:536)
                                                   at 
android.support.v7.view.WindowCallbackWrapper.onPreparePanel(WindowCallbackWrapper.java:93)
                                                   at 
android.support.v7.app.AppCompatDelegateImplBase$AppCompatWindowCallbackBase.onPreparePanel(AppCompatDelegateImplBase.java:297)
                                                   at 
android.support.v7.view.WindowCallbackWrapper.onPreparePanel(WindowCallbackWrapper.java:93)
                                                   at 
android.support.v7.app.ToolbarActionBar$ToolbarCallbackWrapper.onPreparePanel(ToolbarActionBar.java:567)
                                                   at 
android.support.v7.app.ToolbarActionBar.populateOptionsMenu(ToolbarActionBar.java:455)
                                                   at 
android.support.v7.app.ToolbarActionBar$1.run(ToolbarActionBar.java:61)
                                                   at 
android.os.Handler.handleCallback(Handler.java:733)
                                                   at 
android.os.Handler.dispatchMessage(Handler.java:95)
                                                   at 
android.os.Looper.loop(Looper.java:146)
                                                   at 
android.app.ActivityThread.main(ActivityThread.java:5602)
                                                   at 
java.lang.reflect.Method.invokeNative(Native Method)
                                                   at 
java.lang.reflect.Method.invoke(Method.java:515)
                                                   at 
com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1283)
                                                   at 
com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1099)
                                                   at 
dalvik.system.NativeStart.main(Native Method)
12-01 03:22:19.915 16257-16500/? D/ACRA: Creating DialogIntent for 
1512094938000.stacktrace exception=java.lang.IllegalStateException: 
Couldn't read row 0, col 0 from CursorWindow.  Make sure the Cursor is 
initialized correctly before accessing data from it.


Le 30/11/2017 à 23:09, Tim Rae a écrit :

**Comment by timrae** [comment]
Well this is indeed the same exception that we were seeing before with the cursor size limit issue (#4033). I don't really have any concrete ideas, so probably we'll have to just keep removing stuff from the collection until we find out what it's tripping on. For example, does the error go away if you delete *all* the models for example? All the decks?

**Comment by Arthur-Milchior** [comment]
First: thank you for taking time on this bug. I think I did forgot to 
mention it earlier. Sorry.


Here are the experiences I did.

Removing the last half of the content of «models»  and every cards and 
every nodes: everything works fine (apart from the fact that my 
collection is now empty). Removing the first half leads to the same result.

It is however pretty strange, since even before removing anything, 
models was under 20MB.  I assume that, in fact, 20MB is the row limit 
and not the string limit. This is coherent with the following experience:

Removing all decks and dconf apart from the default one led to the 
disparition of the old message.

However, now, a new problem did appear. I only see the vague message 
stating that there is a problem in the database. In french, it is: «La 
collection n'est pas ouverte. Une erreur s'est produite en accédant à 
votre collection... » (I assume you can find the translation if you know 
where the list of sentences is saved)

In each preceding experiments, I moved collection.anki2 from my desktop 
to the ankidroid on my android.


My current solution now is to use ankiweb. But it is really not as nice.

One question: since you were able to reproduce my bug, why do you need 
me to do all of those things ? Is it only because it is my bug and you 
don't want to spend too much time on it ? Because I really don't see 
what I add doing the manipulation. This is especially a problem for me 
because I don't know android development, so it usually take me hours to 
follow your instructions. I think most time is lost by trying to 
understand exactly what should be done, and not by doing the actual 
experiments.


Thank you again,

Arthur Milchior


Le 01/12/2017 à 09:50, Tim Rae a écrit :

**Comment by timrae** [comment]
Actually I wasn't able to reproduce it exactly, I got a coredump not an illegal state exception... and yes I'm super busy and don't have much time to spend on AnkiDroid these days.

You're probably right that it's the row limit rather than the cell limit. So the workaround is to either delete some more models if you can, or split your collection into several parts using profiles. Note that profiles are not strictly supported by AnkiDroid, but you can easily compile your own "parallel" versions by following the instructions on the wiki to achieve this.

**Comment by timrae** [comment]
As for fixing the error message, I don't really have time to do this right now, but you should now have an Android development environment setup, so it shouldn't be too much work if you were interested in making a merge request.

**Comment by timrae** [comment]
I haven't tested this yet, but I think this is the magic number here, which needs to be less than 1048576:

`sqlite3 collection.anki2 "length(hex(id))+length(hex(crt))+length(hex(mod))+length(hex(scm))+length(hex(ver))+length(hex(dty))+length(hex(usn))+length(hex(ls))+length(hex(conf))+length(hex(models))+length(hex(decks))+length(hex(dconf))+length(hex(tags)) as 'total' from col"`

**Comment by mikehardy** [comment]
This should be fixed in the lastest alpha release? https://github.com/ankidroid/Anki-Android/releases

**Comment by Arthur-Milchior** [comment]
@mikehardy Thank you so much !

**Comment by mikehardy** [comment]
@Arthur-Milchior - Only replying to say that @dae deserves all credit! Smooth idea he used to get around this
